src/filesystem/impls/filer/lib/handlers.js

```diff
@@ -4,72 +4,26 @@ define(function (require, exports, module) {
     var CSSRewriter = require("filesystem/impls/filer/lib/CSSRewriter");
-    var Filer = require("filesystem/impls/filer/BracketsFiler");
-    var Path  = Filer.Path;
-
-    function _handle404(err, path, callback) {
-        var html = '<!DOCTYPE html>' +
-                   '<html><head>' +
-                   '<title>404 Not Found</title>' +
-                   '</head><body>' +
-                   '<h1>Not Found</h1>' +
-                   '<p>The requested URL ' + path + ' was not found on this server.</p>' +
-                   '<hr>' +
-                   '<address>nohost/0.0.1 (Web) Server</address>' +
-                   '</body></html>';
-        callback(err, Content.toURL(html, 'text/html'));
-    }
-
-    function _handleHTML(path, content, mime, callback) {
-        HTMLRewriter.rewrite(path, content, function(err, html) {
-            if(err) {
-                Log.error('unable to read html for `' + path + '`');
-                // TODO: best way to deal with error here? 500?
-                return _handle404(err, path, callback);
-            }
-
-            callback(null, Content.toURL(html, mime));
-        });
-    }
-
-    function _handleCSS(path, content, mime, callback) {
-        CSSRewriter.rewrite(path, content, function(err, css) {
-            if(err) {
-                Log.error('unable to read css for `' + path + '`');
-                // TODO: best way to deal with error here? 500?
-                return _handle404(err, path, callback);
-            }
-
-            callback(null, Content.toURL(css, mime));
-        });
-    }
+    var Path  = require("filesystem/impls/filer/BracketsFiler").Path;
 
     /**
      * Send the raw file, making it somewhat more readable
      */
-    function handleFile(path, callback) {
-        var fs = Filer.fs();
+    function handleFile(path, data) {
         var ext = Path.extname(path);
-        var encoding = Content.isUTF8Encoded(ext) && 'utf8';
         var mimeType = Content.mimeFromExt(ext);
 
-        fs.readFile(path, encoding, function(err, data) {
-            if(err) {
-                Log.error('unable to read `' + path + '` in handler');
-                return _handle404(err, path, callback);
-            }
+        if(Content.isHTML(ext)) {
+            // We call toString on the data so that only utf8 data is used if a
+            // buffer was passed in as a parameter
+            data = HTMLRewriter.rewrite(path, data.toString());
+        } else if(Content.isCSS(ext)) {
+            data = CSSRewriter.rewrite(path, data.toString());
+        }
 
-            if(Content.isHTML(ext)) {
-                _handleHTML(path, data, mimeType, callback);
-            } else if(Content.isCSS(ext)) {
-                _handleCSS(path, data, mimeType, callback);
-            } else {
-                callback(null, Content.toURL(data, mimeType));
-            }
-        });
+        return Content.toURL(data, mimeType);
```

Comment:
In your follow up patch, we should just cache this path/data in blob utils so it gets auto-revoked on follow-up generations (e.g., reload).